core/src/main/java/org/apache/iceberg/InternalData.java

```diff
@@ -0,0 +1,159 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.common.DynMethods;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InternalData {
+  private InternalData() {}
+
+  private static final Logger LOG = LoggerFactory.getLogger(InternalData.class);
+  private static final Map<FileFormat, Function<OutputFile, WriteBuilder>> WRITE_BUILDERS =
+      Maps.newConcurrentMap();
+  private static final Map<FileFormat, Function<InputFile, ReadBuilder>> READ_BUILDERS =
+      Maps.newConcurrentMap();
+
+  static {
+    Avro.register();
+
+    try {
+      DynMethods.StaticMethod registerParquet =
+          DynMethods.builder("register")
+              .impl("org.apache.iceberg.parquet.Parquet")
+              .buildStaticChecked();
+
+      registerParquet.invoke();
+
+    } catch (NoSuchMethodException e) {
+      // failing to load Parquet is normal and does not require a stack trace
+      LOG.info("Unable to register Parquet for metadata files: {}", e.getMessage());
+    }
+  }
+
+  public static void register(
+      FileFormat format,
+      Function<OutputFile, WriteBuilder> writeBuilder,
+      Function<InputFile, ReadBuilder> readBuilder) {
+    WRITE_BUILDERS.put(format, writeBuilder);
+    READ_BUILDERS.put(format, readBuilder);
+  }
+
+  public static WriteBuilder write(FileFormat format, OutputFile file) {
+    Function<OutputFile, WriteBuilder> writeBuilder = WRITE_BUILDERS.get(format);
+    if (writeBuilder != null) {
+      return writeBuilder.apply(file);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot write using unregistered internal data format: " + format);
+  }
+
+  public static ReadBuilder read(FileFormat format, InputFile file) {
+    Function<InputFile, ReadBuilder> readBuilder = READ_BUILDERS.get(format);
+    if (readBuilder != null) {
+      return readBuilder.apply(file);
+    }
+
+    throw new UnsupportedOperationException(
+        "Cannot read using unregistered internal data format: " + format);
+  }
+
+  public interface WriteBuilder {
+    /** Set the file schema. */
+    WriteBuilder schema(Schema schema);
+
+    /** Set the file schema's root name. */
+    WriteBuilder named(String name);
+
+    /**
+     * Set a writer configuration property.
+     *
+     * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
```

Comment:
```suggestion
     * <p>Write configuration affects this writer's behavior. To add metadata properties to the written file use  {@link
```
?

Reply:
"This" doesn't refer to a writer. This is configuring a builder that creates the writer, so I think that the existing language is correct.

Reply:
This is alignment with the existing writer APIs. We have this for quite some time already, so I think we should stick to it unless we have strong reasons to change. The change in the javadoc helps too